lib/grpc/client/adapters/mint/connection_process/connection_process.ex

```diff
@@ -0,0 +1,390 @@
+    send_eof? = from == nil
+
+    case stream_body(state.conn, request_ref, body, send_eof?) do
+      {:ok, conn} ->
+        if from != nil do
+          GenServer.reply(from, :ok)
+        end
+
+        check_request_stream_queue(State.update_conn(state, conn))
+
+      {:error, conn, error} ->
+        if from != nil do
+          GenServer.reply(from, {:error, error})
+        else
+          state
+          |> State.stream_response_pid(request_ref)
+          |> StreamResponseProcess.consume(:error, error)
+        end
+
+        check_request_stream_queue(State.update_conn(state, conn))
+    end
+  end
+
+  defp stream_body(conn, request_ref, body, true = _stream_eof?) do
+    with {:ok, conn} <- Mint.HTTP.stream_request_body(conn, request_ref, body),
+         {:ok, conn} <- Mint.HTTP.stream_request_body(conn, request_ref, :eof) do
+      {:ok, conn}
+    end
+  end
+
+  defp stream_body(conn, request_ref, body, false = _stream_eof?) do
+    with {:ok, conn} <- Mint.HTTP.stream_request_body(conn, request_ref, body) do
+      {:ok, conn}
+    end
+  end
+
+  def check_request_stream_queue(state) do
+    if :queue.is_empty(state.request_stream_queue) do
+      {:noreply, state}
+    else
+      {:noreply, state, {:continue, :process_request_stream_queue}}
+    end
+  end
+
+  defp chunk_body(body, bytes_length) do
+    <<head::binary-size(bytes_length), tail::binary>> = body
+    {head, tail}
+  end
+
+  def get_window_size(conn, ref) do
+    min(
+      Mint.HTTP2.get_window_size(conn, {:request, ref}),
+      Mint.HTTP2.get_window_size(conn, :connection)
+    )
+  end
+
+  defp finish_all_pending_requests(state) do
+    new_state =
+      state.request_stream_queue
+      |> :queue.to_list()
+      |> Enum.reduce(state, fn request, acc_state ->
+        case request do
+          {ref, _body, nil} ->
+            acc_state
+            |> State.stream_response_pid(ref)
+            |> send_connection_close_and_end_stream_response()
+
+          {ref, _body, from} ->
+            acc_state
+            |> State.stream_response_pid(ref)
+            |> send_connection_close_and_end_stream_response()
+
+            GenServer.reply(from, {:error, @connection_closed_error})
+        end
+
+        {ref, _, _} = request
+        {_ref, new_state} = State.pop_ref(acc_state, ref)
+
+        new_state
+      end)
+
+    # Inform the parent that the connection is down
+    send(new_state.parent, {:elixir_grpc, :connection_down, self()})
+
+    new_state.requests
+    |> Map.keys()
+    |> Enum.each(fn ref ->
+      new_state
+      |> State.stream_response_pid(ref)
+      |> send_connection_close_and_end_stream_response()
+    end)
+
+    {:noreply, State.update_request_stream_queue(%{new_state | requests: %{}}, :queue.new())}
+  end
+
+  defp send_connection_close_and_end_stream_response(pid) do
+    StreamResponseProcess.consume(pid, :error, @connection_closed_error)
+    StreamResponseProcess.done(pid)
```

Comment:
I think a match on the consume result would be nice, to ensure that it  succeeded

Reply:
This function calls `GenServer.cast`, always returns `:ok`.

Reply:
Yeah, but I think that GenServer.cast should be avoided either way, if we can

Reply:
https://github.com/elixir-grpc/grpc/pull/272#discussion_r1057331090 I'll reply in this thread

Reply:
Mark this one as fixed.